File: hackerearth/Circuit March 2017/xor-rectangle.cpp
#include <bits/stdc++.h>

using namespace std;

const int N = 1e6 + 7;

int bit[20][N];
int a[N], n, q, xl, xr, yl, yr;

int main() {
  scanf("%d", &n);
  for (int i = 0; i < n; i++) {
    scanf("%d", a+i);
    for (int j = 0; j < 20; j++)
      bit[j][i+1] = bit[j][i] + ((a[i] & (1<<j)) ? 1 : 0);
  }
  scanf("%d", &q);
  while (q--) {
    scanf("%d %d %d %d", &xl, &yl, &xr, &yr);
    xl--; yl--;
    long long ans = 0;
    for (int i = 0; i < 20; i++) {
      long long onex = bit[i][xr]-bit[i][xl], oney = bit[i][yr]-bit[i][yl],
      zerox = xr-xl-onex, zeroy = yr-yl-oney;
      ans += (onex * zeroy + oney * zerox) << i;
    }
    printf("%lld\n", ans);
  }

  return 0;
}

File: hackerearth/march-circuits-17/xor-rectangle.py
#/bin/python3

def gets():
    return list(map(int, input().strip().split()))
n = int(input())
s = gets()
bit = [[0] + [1 if (s[j] & (1<<i)) else 0 for j in range(n)] for i in range(20)]
for i in range(20):
    for j in range(n):
        bit[i][j+1] += bit[i][j]
q = int(input())
for _ in range(q):
    xl, yl, xr, yr = gets()
    ans = 0
    for j in range(20):
        onex, oney = bit[j][xr] - bit[j][xl-1], bit[j][yr] - bit[j][yl-1]
        zerox, zeroy = xr-xl+1-onex, yr-yl+1-oney
        ans += (onex * zeroy + oney * zerox) << j
    print(ans)
